File: quant1x/id/id128/option.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

from __future__ import annotations

from typing import Callable, TYPE_CHECKING

if TYPE_CHECKING:
    from .hlc import HLC

Option = Callable[["HLC"], None]


def with_clock(now: Callable[[], int]) -> Option:
    def apply(hlc: "HLC") -> None:
        if now is not None:
            hlc._now = now

    return apply


def with_logical_seed(seed: int) -> Option:
    def apply(hlc: "HLC") -> None:
        hlc.seed = seed & 0xFFFF

    return apply


def with_state_file(path: str) -> Option:
    def apply(hlc: "HLC") -> None:
        if path:
            from .state_store import FileStateStore

            hlc.store = FileStateStore(path, sync_every=hlc.sync_every)

    return apply


def with_state_sync_every(every: int) -> Option:
    def apply(hlc: "HLC") -> None:
        hlc.sync_every = max(1, int(every))
        if hlc.store is not None:
            hlc.store.sync_every = hlc.sync_every

    return apply

File: quant1x/id/id128/Option.java
// Copyright (c) 2026 Quant1X. All rights reserved.
// Author: wangfeng <wangfengxy@sina.cn>
// SPDX-License-Identifier: MIT

package quant1x.id.id128;

import java.util.function.LongSupplier;

/**
 * HLC 构造期配置项，对应 Go 版的函数式 Option 与 Python 版的
 * {@code with_clock/with_logical_seed/with_state_file/with_state_sync_every}。
 *
 * <pre>
 * HLC hlc = new HLC(
 *         Option.withClock(System::currentTimeMillis),
 *         Option.withLogicalSeed(7),
 *         Option.withStateFile("id.state"),
 *         Option.withStateSyncEvery(1));
 * </pre>
 */
@FunctionalInterface
public interface Option {

    /**
     * 将配置应用到 {@link HLC} 实例（仅构造期调用）。
     */
    void apply(HLC hlc);

    /**
     * 注入时钟源（返回当前物理时间，毫秒）。
     * 缺省为 {@link System#currentTimeMillis()}。
     *
     * @param now 时钟源，为 null 时忽略
     */
    static Option withClock(LongSupplier now) {
        return hlc -> {
            if (now != null) {
                hlc.now = now;
            }
        };
    }

    /**
     * 注入逻辑种子（16 位），用于回退场景的随机化。
     * 缺省为加密随机数。
     *
     * @param seed 16 位种子
     */
    static Option withLogicalSeed(int seed) {
        return hlc -> hlc.seed = seed & 0xFFFF;
    }

    /**
     * 启用基于文件的状态持久化（跨重启强唯一）。
     *
     * @param path 状态文件路径，为 null 或空时忽略
     */
    static Option withStateFile(String path) {
        return hlc -> {
            if (path != null && !path.isEmpty()) {
                hlc.store = new FileStateStore(path);
            }
        };
    }

    /**
     * 设置每次 fsync 之间最多追加的状态记录条数。
     * 缺省为 1；大于 0。
     *
     * @param every 每次 fsync 之间最多追加的记录条数
     */
    static Option withStateSyncEvery(long every) {
        return hlc -> {
            hlc.syncEvery = Math.max(1, every);
            if (hlc.store instanceof FileStateStore) {
                ((FileStateStore) hlc.store).syncEvery = hlc.syncEvery;
            }
        };
    }
}
